xbmc/platform/android/activity/XBMCApp.cpp

```diff
@@ -1621,3 +1624,41 @@ void CXBMCApp::surfaceDestroyed(CJNISurfaceHolder holder)
 
   m_window.reset();
 }
+
+void CXBMCApp::OnVoiceRecognition()
+{
+  if (CJNISpeechRecognizer::isRecognitionAvailable(CXBMCApp::Get()))
+  {
+    if (CJNIContext::checkCallingOrSelfPermission("android.permission.RECORD_AUDIO") == CJNIPackageManager::PERMISSION_GRANTED) 
+    {
+      // must be called from the main thread
+      runNativeOnUiThread(StartRecognition, nullptr);
+    }
+    else
+    {
+      CLog::Log(LOGDEBUG, "CXBMCApp OnVoiceRecognition(): permission RECORD_AUDIO is not allowed");
+      std::string error = "Allow Microphone permission to Kodi app in Android Settings";
+      CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, "Speech to text", error);
+    }
+  }
+  else
+  {
+    CLog::Log(LOGDEBUG, "CXBMCApp OnVoiceRecognition(): Google speech service no available");
+    CGUIDialogKaiToast::QueueNotification(CGUIDialogKaiToast::Info, "Speech to text", "Google speech service no available");
```

Comment:
string localization

Reply:
To get rid of the gui dependency here, OnVoiceRecognition should return an error code (an enum value) and the calling GUI code should raise the toast.